### ISSUE TITLE
DMTO: submission 1.0 failed to parse

### ISSUE THREAD

**Issue by jvendetti** [ncbo/bioportal-project#16]
Version 1.0 of the [Diabetes Mellitus Treatment Ontology](http://bioportal.bioontology.org/ontologies/DMTO) (DMTO) failed to parse.

Pertinent snippet from the parsing log file:

```
2017-03-30T16:23:53 [main] INFO  o.s.n.o.OntologyParserCommand - Parse result: false

2017-03-30T16:23:53 [main] INFO  o.s.n.o.OntologyParserCommand - Output triples in: {}/srv/ncbo/repository/DMTO/2/owlapi.xrdf

2017-03-30T16:23:53 [main] INFO  o.s.n.o.OntologyParserCommand - Finished parsing!

I, [2017-03-30T16:23:53.092004 #28987]  INFO -- : OWLAPI Java command: parsing finished successfully.
E, [2017-03-30T16:23:53.092323 #28987] ERROR -- : LinkedData::Parser::OWLAPIParserException: OWLAPI java command exited with  0. Output file /srv/ncbo/repository/DMTO/2/owlapi.xrdf cannot be found.
/srv/ncbo/ncbo_cron/vendor/bundle/ruby/2.3.0/bundler/gems/ontologies_linked_data-9fb051182d58/lib/ontologies_linked_data/parser/owlapi.rb:125:in `block in call_owlapi_java_command'
        /usr/local/rbenv/versions/2.3.3/lib/ruby/2.3.0/open3.rb:205:in `popen_run'
        /usr/local/rbenv/versions/2.3.3/lib/ruby/2.3.0/open3.rb:95:in `popen3'
        /srv/ncbo/ncbo_cron/vendor/bundle/ruby/2.3.0/bundler/gems/ontologies_linked_data-9fb051182d58/lib/ontologies_linked_data/parser/owlapi.rb:76:in `call_owlapi_java_command'
        /srv/ncbo/ncbo_cron/vendor/bundle/ruby/2.3.0/bundler/gems/ontologies_linked_data-9fb051182d58/lib/ontologies_linked_data/parser/owlapi.rb:153:in `parse'
        /srv/ncbo/ncbo_cron/vendor/bundle/ruby/2.3.0/bundler/gems/ontologies_linked_data-9fb051182d58/lib/ontologies_linked_data/models/ontology_submission.rb:430:in `generate_rdf'
        /srv/ncbo/ncbo_cron/vendor/bundle/ruby/2.3.0/bundler/gems/ontologies_linked_data-9fb051182d58/lib/ontologies_linked_data/models/ontology_submission.rb:860:in `process_submission'
        /srv/ncbo/ncbo_cron/lib/ncbo_cron/ontology_submission_parser.rb:178:in `process_submission'
        ./bin/ncbo_ontology_process:98:in `block in <main>'
        ./bin/ncbo_ontology_process:81:in `each'
        ./bin/ncbo_ontology_process:81:in `<main>'
```

**Comment by mdorf** [comment]
The underlying error, when run through an OWL API Wrapper test appears to be:

```
Caused by: org.semanticweb.owlapi.rdf.rdfxml.renderer.IllegalElementNameException: 
Illegal Element Name (Element Is Not A QName): 
https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000081
https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0001649
http://www.co-ode.org/ontologies/ont.owl#DDO:0000002
https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000051
https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000909
https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000079
```

**Comment by jvendetti** [comment]
Attempted to open this ontology in Protege (version 5.2.0).  Opens without any parsing errors in the log.  The UI does indicate an error for the dc:date annotation:

![screenshot 2017-04-14 17 53 41](https://cloud.githubusercontent.com/assets/1696923/25059429/764a2aac-213b-11e7-8c48-fef1e1a6e910.png)

This should be fixed, but it's not clear if it would help with the BioPortal parsing process.

**Comment by graybeal** [comment]
Just recording a data point, as I was trying to find these throughout.

I notice all of them are declared as Annotation Properties in DMTO.owl at lines 24552 etc., e.g.:

```
    <Declaration>
        <AnnotationProperty IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000051"/>
    </Declaration>
```
and also as IRIs around line 163469, e.g.:
```
    <AnnotationAssertion>
        <AnnotationProperty abbreviatedIRI="rdfs:label"/>
        <IRI>https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000051</IRI>
        <Literal xml:lang="en" datatypeIRI="&rdf;PlainLiteral">imported from</Literal>
    </AnnotationAssertion>
```

Not seeing them used elsewhere.

**Comment by jvendetti** [comment]
Yes, that's correct.  An exception is thrown from [line 617 in the owlapi_wrapper code](https://github.com/ncbo/owlapi_wrapper/blob/master/src/main/java/org/stanford/ncbo/oapiwrapper/OntologyParser.java#L617) when we attempt to serialize the ontology file to disk in RDF/XML format.  The detailed message from the exception reports the "Element Is Not A QName" error on six entities, all of which are declared as annotation properties in the original ontology source file.

**Comment by jvendetti** [comment]
Yes, that's correct.  An exception is thrown from [line 617 in the owlapi_wrapper code](https://github.com/ncbo/owlapi_wrapper/blob/master/src/main/java/org/stanford/ncbo/oapiwrapper/OntologyParser.java#L617) when we attempt to serialize the ontology file to disk in RDF/XML format.  The detailed message from the exception reports the "Element Is Not A QName" error on six entities, all of which are declared as annotation properties in the original ontology source file.

**Comment by vemonet** [comment]
I experienced a similar problem some months ago. The problem comes from the RDFXMLDocumentFormat
https://github.com/ncbo/owlapi_wrapper/blob/master/src/main/java/org/stanford/ncbo/oapiwrapper/OntologyParser.java#L615

If you change this line to save the ontology as OWLXMLDocumentFormat, NQuadsDocumentFormat or RioTurtleDocumentFormat it will work fine (excepting for the OWL_IMPORT_MISSING of http://www.owl-ontologies.com/OntoFood.owl, but this is not the problem here)
So the problem comes from the RDFXMLDocumentFormat serialization

In my case I found out that the problems came from a bad relation description in the OBO file (it was for a OBO file) : 
```
[Typedef]
id: 12
name: causes disease
comment: Argument 1 = bacteria - argument 2 = disease.
```
Using integer as ID was causing trouble
```
[Typedef]
id: causes_disease
name: causes disease
comment: Argument 1 = bacteria - argument 2 = disease.
```
Was the right way to do it

**Comment by vemonet** [comment]
In our case we can see that the 6 URI that caused the error have a thing in common in comparison to the other AnnotationProperty Declarations: they are defined using IRI and not AbbreviatedIRI

Example (the first is good, the second cause the error):
```xml
<Declaration>
  <AnnotationProperty abbreviatedIRI="foaf:mbox"/>
</Declaration>
<Declaration>
  <AnnotationProperty IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000051"/>
</Declaration>
```

**Comment by vemonet** [comment]
I fixed the DMTO.owl file by adding 2 prefixes:

```xml
    <Prefix name="coode" IRI="http://www.co-ode.org/ontologies/ont.owl#"/>
    <Prefix name="dmtobp" IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#"/>
```

And changing the AnnotationProperty Declarations (around lines 24 539)

```xml
    <Declaration>
        <AnnotationProperty abbreviatedIRI="coode:DDO:0000002"/>
    </Declaration>
    <Declaration>
        <AnnotationProperty abbreviatedIRI="dmtobp:DMTO:0000051"/>
    </Declaration>
    <Declaration>
        <AnnotationProperty abbreviatedIRI="dmtobp:DMTO:0000079"/>
    </Declaration>
    <Declaration>
        <AnnotationProperty abbreviatedIRI="dmtobp:DMTO:0000081"/>
    </Declaration>
    <Declaration>
        <AnnotationProperty abbreviatedIRI="dmtobp:DMTO.owl#DMTO:0000909"/>
    </Declaration>
    <Declaration>
        <AnnotationProperty abbreviatedIRI="dmtobp:DMTO.owl#DMTO:0001649"/>
    </Declaration>
```

Now the parsing is working fine (except for the MISSING_IMPORT)

Here is the fixed file (I changed the extension file to txt to be accepted by GitHub)
[DMTO.txt](https://github.com/ncbo/bioportal-project/files/972042/DMTO.txt)

It seems like the RDFXMLDocumentFormat serialization has some special restrictions on AnnotationProperty that are not present in the OWLXMLDocumentFormat for example

**Comment by graybeal** [comment]
I thought the main explanation/option might be to replace IRI with abbreviatedIRI. Will sit with you to discuss the other bits, so that I understand them.

**Comment by vemonet** [comment]
Sorry my first solution was not working in the end.

Here is an uploaded version working: http://stageportal.lirmm.fr/ontologies/DBTST

If found that the problem comes from the IRI design. It is not valid because of the `:`

* http://www.co-ode.org/ontologies/ont.owl#DDO:0000002 is not good.
But http://www.co-ode.org/ontologies/ont.owl#DDO_0000002 is good.
We can see that usually the co-ode URI are usually designed like this: http://www.co-ode.org/ontologies/ont.owl#DDO_0010000. So the http://www.co-ode.org/ontologies/ont.owl#DDO:0000002 may be a typo

* It is the same for the DMTO IRIs:
```xml
    <Declaration>
        <AnnotationProperty IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000079"/>
    </Declaration>
    <AnnotationAssertion>
        <AnnotationProperty abbreviatedIRI="rdfs:label"/>
        <IRI>https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000079</IRI>
        <Literal xml:lang="en" datatypeIRI="&rdf;PlainLiteral">RxNorm name</Literal>
    </AnnotationAssertion>
```
Is not valid.

We need to change it as 
```xml
    <Declaration>
        <AnnotationProperty IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO_0000079"/>
    </Declaration>
    <AnnotationAssertion>
        <AnnotationProperty abbreviatedIRI="rdfs:label"/>
        <IRI>https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO_0000079</IRI>
        <Literal xml:lang="en" datatypeIRI="&rdf;PlainLiteral">RxNorm name</Literal>
    </AnnotationAssertion>
```

* Notice that we don't get any error with Class that have IRI with `:`. So the 
```xml
    <Declaration>
        <Class IRI="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000053"/>
    </Declaration>
    <AnnotationAssertion>
        <AnnotationProperty abbreviatedIRI="rdfs:label"/>
        <IRI>https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000053</IRI>
        <Literal xml:lang="en" datatypeIRI="&rdf;PlainLiteral">orderable drug form</Literal>
    </AnnotationAssertion>
```

It is normal since in the resulting owlapi.xrdf file we can see that the class URIs (with `:`) are only used as full IRI, like this: `<Class rdf:about="https://bioportal.bioontology.org/ontologies/DMTO.owl#DMTO:0000012">`

But  the AnnotationProperty are used once with prefix:
```xml
    <Class rdf:about="http://purl.obolibrary.org/obo/DDO.owl#DDO_0009287">
        <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/DDO.owl#DDO_0009279"/>
        <metadata:prefixIRI rdf:datatype="http://www.w3.org/2001/XMLSchema#string">DDO:DDO_0009287</metadata:prefixIRI>
        <DDO:DDO_0000001 rdf:datatype="http://www.w3.org/2001/XMLSchema#integer">45007003</DDO:DDO_0000001>
        <rdfs:label>hypotension</rdfs:label>
        <DMTO:DMTO_0001649>arterial hypotension</DMTO:DMTO_0001649>
        <DMTO:DMTO_0001649>hypopiesis</DMTO:DMTO_0001649>
        <DMTO:DMTO_0001649>low blood pressure</DMTO:DMTO_0001649>
    </Class>
```

**Comment by graybeal** [comment]
The problem has been determined to be due to the fact these 6 strings have been declared as IRIs, but colon is not a legal character in that context. (Colon is legal in Abbreviated IRIs, which everything else is.)  So probably these should be Abbreviated IRIs.

Original requestor has been notified and will be making changes to the ontology.

**Comment by jvendetti** [comment]
User uploaded a new version today.  Ontology still did not parse because he missed correcting some of the invalid IRIs for annotation properties with the colon characters in them.  I corrected the remaining 3 instances, and uploaded a new ZIP file.  

The new submission parsed successfully, but there's still the matter of the missing import for the OntoFood.owl file.  He included the ontology file in the last ZIP file he submitted, and we've made claims on the support list that BioPortal should find and process all ontologies in the submission directory.  However, I found a bug in the code in the [owlapi_wrapper](https://github.com/ncbo/owlapi_wrapper) project that prevents this functionality from working.  The current state of the system is that we'll only process imported ontologies if the import statements in the master files are correct and can be resolved to an actual file.

As a workaround, I used the Protege editing tool to merge all of his ontologies into a single file.  I uploaded another new version with this single file and the ontology is now fully available / parsed in BioPortal.  I notified the user with regard to the workaround we used.
